dohclient.go

```diff
@@ -275,66 +273,113 @@ func dohQuicTransport(endpoint string, opt DoHClientOptions) (http.RoundTripper,
 		},
 		Dial: dialer,
 	}
-	return &http3ReliableRoundTripper{tr, pool}, nil
+	return tr, nil
 }
 
-func quicDial(hostname, rAddr string, lAddr net.IP, tlsConfig *tls.Config, config *quic.Config, pool *udpConnPool) (quic.EarlyConnection, error) {
-	udpAddr, err := net.ResolveUDPAddr("udp", rAddr)
-	if err != nil {
-		return nil, err
-	}
-	udpConn, err := net.ListenUDP("udp", &net.UDPAddr{IP: lAddr, Port: 0})
+// QUIC connection that automatically restarts when it's used after having timed out. Needed
+// since the quic-go RoundTripper doesn't have any connection management and timed out
+// connection aren't restarted. This one doesn't support Early connection, and instead just
+// uses a regular connection.
+type quicConnection struct {
+	quic.Connection
+
+	hostname  string
+	rAddr     string
+	lAddr     net.IP
+	tlsConfig *tls.Config
+	config    *quic.Config
+	mu        sync.Mutex
+	udpConn   *net.UDPConn
+
+	expiredContext context.Context
+}
+
+func newQuicConnection(hostname, rAddr string, lAddr net.IP, tlsConfig *tls.Config, config *quic.Config) (quic.EarlyConnection, error) {
+	connection, udpConn, err := quicDial(hostname, rAddr, lAddr, tlsConfig, config)
 	if err != nil {
 		return nil, err
 	}
-	pool.add(udpConn)
-	return quic.DialEarly(udpConn, udpAddr, hostname, tlsConfig, config)
+	expired, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	return &quicConnection{
+		hostname:       hostname,
+		rAddr:          rAddr,
+		lAddr:          lAddr,
+		tlsConfig:      tlsConfig,
+		config:         config,
+		udpConn:        udpConn,
+		Connection:     connection,
+		expiredContext: expired,
+	}, nil
 }
 
-// Wrapper for http3.RoundTripper due to https://github.com/quic-go/quic-go/issues/765
-// This wrapper will transparently re-open expired connections. Should be removed once the issue
-// has been fixed upstream.
-type http3ReliableRoundTripper struct {
-	*http3.RoundTripper
-	pool *udpConnPool
+func (s *quicConnection) HandshakeComplete() context.Context {
+	return s.expiredContext
 }
 
-func (r *http3ReliableRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
-	resp, err := r.RoundTripper.RoundTrip(req)
-	if netErr, ok := err.(net.Error); ok && (netErr.Timeout() || netErr.Temporary()) {
-		r.pool.closeAll()
-		r.RoundTripper.Close()
-		resp, err = r.RoundTripper.RoundTrip(req)
+func (s *quicConnection) OpenStreamSync(ctx context.Context) (quic.Stream, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	stream, err := s.Connection.OpenStreamSync(ctx)
+	if err != nil {
+		// Try to open a new connection, but clean up our mess before we do so
+		_ = s.Connection.CloseWithError(DOQNoError, "")
+		// and then we need to close the connection / socket ourselves as we own
+		// the UDP socket not quic-go
+		// c.f. https://github.com/quic-go/quic-go/issues/1457
+		_ = s.udpConn.Close()
+		s.udpConn = nil
```

Comment:
This could be a problem when the next attempt fails as well since there's going to be a nil-ptr panic above. We could either make sure it's never nil, or add a condition

Reply:
Yeah, I'll put a guard around it like above.